### PR TITLE
Create pid file to avoid having two daemons.

### PR DESCRIPTION
**Pull request by jjnicola**
Adds the --pid-file option in the parser.
Add method to create and remove the pid file.
Set the signal handler and atexit to remove the pid file when the daemon finishes.

How to:
by default, the pid path is /run/ospd/ospd.py. You will have to create the directory and set the right permissions.
```
sudo mkdir /run/ospd`
sudo chown jnicola:jnicola /run/ospd
```

Or you can run ospd-daemon with the option --pid-file and use an existing directory with rw permissions:

```
ospd-openvas --pid-file /tmp/ospd-openvas.pid
```

This is also valid to do it with the config file, adding under the right section the option. E.g.:

```
[OSPD - openvas]
log_level = DEBUG
socket_mode = 0o770
unix_socket = /tmp/openvas.sock
pid_file = /tmp/openvas.sock
```

### CODE REVIEW

File: ospd/misc.py [hunk 1]
```diff
@@ -408,3 +408,31 @@ def go_to_background():
...
         LOGGER.error('Fork failed: %s', errmsg)
         sys.exit(1)
+
+def create_pid(pidfile):
+    """ Check if there is an already running daemon and creates the pid file.
+    Otherwise gives an error. """
+
+    pid = str(os.getpid())
+
+    if os.path.isfile(pidfile):
```
Comment (bjoernricks) [inline ospd/misc.py]: Please use [pathlib](https://docs.python.org/3/library/pathlib.html) instead
```suggestion
    if Path(pidfile).is_file():
```

File: ospd/misc.py [hunk 2]
```diff
@@ -408,3 +408,31 @@ def go_to_background():
...
+        LOGGER.error("There is an already running process.")
+        return False
+
+    try:
+        f = open(pidfile, 'w')
+        f.write(pid)
+        f.close()
+    except (FileNotFoundError, PermissionError) as e:
```
Comment (bjoernricks) [inline ospd/misc.py]: Either use the with statement (safest solution) `ẁith open(...) as f:` or add a finally statement to always close the file. Currently a different exception could possibly occur and the file would not be closed at all.

File: ospd/misc.py [hunk 3]
```diff
@@ -408,3 +408,31 @@ def go_to_background():
...
+        msg = "Failed to create pid file %s. %s" % (os.path.dirname(pidfile), e)
+        LOGGER.error(msg)
+        return False
+
+    return True
+
+def remove_pidfile(pidfile, signum=None, frame=None):
+    """ Removes the pidfile before ending the daemon. """
+    if os.path.isfile(pidfile):
```
Comment (bjoernricks) [inline ospd/misc.py]: Please use the pathlib here too. Maybe it is even better to already pass pidfile as a Path object to the functions but i am not 100% sure about that.

File: ospd/misc.py [hunk 4]
```diff
@@ -408,3 +408,31 @@ def go_to_background():
...
+        LOGGER.debug("Finishing daemon process")
+        os.remove(pidfile)
```
Comment (bjoernricks) [inline ospd/misc.py]: Should be done via a Path instance 
```python3
pidpath = Path(pidfile)
if pidfile.is_file():
  ...
  pidfile.unlink()
```

File: ospd/main.py [hunk 5]
```diff
@@ -22,10 +22,13 @@
 
 import os
 import sys
+import atexit
+import signal
+from functools import partial
```
Comment (bjoernricks) [inline ospd/main.py]:
```suggestion

from functools import partial
```

File: ospd/misc.py (hunk 2 above)
Comment (jjnicola) [reply ospd/misc.py]: It uses now the with statement.

**Comment by jjnicola** [comment]
Thank you @bjoernricks !

**Pull request merged**